File: Task1.rs
use std::collections::HashMap;
use std::io::{self, Write};

struct Cell {
    address: String,
    product: String,
    quantity: i32,
}

struct Warehouse {
    zones: i32,
    shelves_per_zone: i32,
    sections_per_shelf: i32,
    shelves_per_section: i32,
    total_capacity: i32,
    used_capacity: i32,
    cells: HashMap<String, Cell>,
}

impl Warehouse {
    fn new(z: i32, spz: i32, sec: i32, spl: i32, cap: i32) -> Self {
        Warehouse {
            zones: z,
            shelves_per_zone: spz,
            sections_per_shelf: sec,
            shelves_per_section: spl,
            total_capacity: cap,
            used_capacity: 0,
            cells: HashMap::new(),
        }
    }

    fn address_valid(&self, address: &str) -> bool {
        if address.len() != 5 {
            return false;
        }
        
        let zone = address.chars().nth(0).unwrap();
        if zone < 'A' || zone >= ('A' as u8 + self.zones as u8) as char {
            return false;
        }
        
        let shelf_part = &address[1..3];
        let shelf = match shelf_part.parse::<i32>() {
            Ok(num) => num,
            Err(_) => return false,
        };
        if shelf < 1 || shelf > self.shelves_per_zone {
            return false;
        }
        
        let section = match address.chars().nth(3).unwrap().to_digit(10) {
            Some(num) => num as i32,
            None => return false,
        };
        if section < 1 || section > self.sections_per_shelf {
            return false;
        }
        
        let shelf_num = match address.chars().nth(4).unwrap().to_digit(10) {
            Some(num) => num as i32,
            None => return false,
        };
        if shelf_num < 1 || shelf_num > self.shelves_per_section {
            return false;
        }
        
        true
    }

    fn add(&mut self, product: &str, quantity: i32, address: &str) {
        if !self.address_valid(address) {
            println!("Ошибка: Неверный адрес: {}", address);
            return;
        }

        if quantity > 10 {
            println!("Ошибка: Нельзя добавить более 10 единиц в ячейку");
            return;
        }

        if let Some(cell) = self.cells.get_mut(address) {
            if cell.product != product {
                println!("Ошибка: Ячейка {} уже содержит {}", address, cell.product);
                return;
            }
            if cell.quantity + quantity > 10 {
                println!("Ошибка: Ячейка {} не может содержать более 10 единиц (сейчас: {})", 
                    address, cell.quantity);
                return;
            }
            cell.quantity += quantity;
        } else {
            self.cells.insert(address.to_string(), Cell {
                address: address.to_string(),
                product: product.to_string(),
                quantity,
            });
        }
        self.used_capacity += quantity;
        println!("Добавлено {} единиц {} в {}", quantity, product, address);
    }

    fn remove(&mut self, product: &str, quantity: i32, address: &str) {
        if !self.address_valid(address) {
            println!("Ошибка: Неверный адрес: {}", address);
            return;
        }

        if !self.cells.contains_key(address) {
            println!("Ошибка: Ячейка {} пуста", address);
            return;
        }

        let cell = self.cells.get_mut(address).unwrap();
        if cell.product != product {
            println!("Ошибка: Ячейка {} содержит {}, а не {}", 
                address, cell.product, product);
            return;
        }

        if cell.quantity < quantity {
            println!("Ошибка: Недостаточно {} в ячейке {} (доступно: {})", 
                product, address, cell.quantity);
            return;
        }

        cell.quantity -= quantity;
        self.used_capacity -= quantity;
        println!("Удалено {} единиц {} из {}", quantity, product, address);

        if cell.quantity == 0 {
            self.cells.remove(address);
        }
    }

    fn info(&self) {
        let total_percent = (self.used_capacity as f64 / self.total_capacity as f64) * 100.0;
        println!("Информация о складе:");
        println!("Общая заполненность: {:.1}%", total_percent);

        let mut zone_capacity = HashMap::new();
        let mut zone_used = HashMap::new();
        
        for zone in 'A'..('A' as u8 + self.zones as u8) as char {
            zone_capacity.insert(zone, 
                self.shelves_per_zone * self.sections_per_shelf * self.shelves_per_section * 10);
            zone_used.insert(zone, 0);
        }

        for (_, cell) in &self.cells {
            let zone = cell.address.chars().next().unwrap();
            *zone_used.get_mut(&zone).unwrap() += cell.quantity;
        }

        for zone in 'A'..('A' as u8 + self.zones as u8) as char {
            let percent = (*zone_used.get(&zone).unwrap() as f64 / 
                          *zone_capacity.get(&zone).unwrap() as f64) * 100.0;
            println!("Зона {} заполнена на {:.1}%", zone, percent);
        }

        println!("\nЗанятые ячейки:");
        for (_, cell) in &self.cells {
            println!("{}: {} ({})", cell.address, cell.product, cell.quantity);
        }

        let total_cells = self.zones * self.shelves_per_zone * 
                         self.sections_per_shelf * self.shelves_per_section;
        let empty_cells = total_cells - self.cells.len() as i32;
        println!("\nПустые ячейки: {}", empty_cells);
    }
}

fn main() {
    let mut warehouse = Warehouse::new(1, 10, 7, 4, 2800);
    
    loop {
        print!(">>> ");
        io::stdout().flush().unwrap();
        
        let mut command = String::new();
        io::stdin().read_line(&mut command).unwrap();
        let command = command.trim();
        
        if command.is_empty() {
            continue;
        }

        let parts: Vec<&str> = command.split_whitespace().collect();
        match parts[0] {
            "ADD" => {
                if parts.len() < 4 {
                    println!("Ошибка: Неправильный формат команды ADD. Используйте: ADD <продукт> <количество> <адрес>");
                    continue;
                }
                
                let product = parts[1];
                let quantity = match parts[2].parse::<i32>() {
                    Ok(num) => num,
                    Err(_) => {
                        println!("Ошибка: Неправильный формат количества. Должно быть целое число");
                        continue;
                    }
                };
                let address = parts[3];
                
                warehouse.add(product, quantity, address);
            },
            "REMOVE" => {
                if parts.len() < 4 {
                    println!("Ошибка: Неправильный формат команды REMOVE. Используйте: REMOVE <продукт> <количество> <адрес>");
                    continue;
                }
                
                let product = parts[1];
                let quantity = match parts[2].parse::<i32>() {
                    Ok(num) => num,
                    Err(_) => {
                        println!("Ошибка: Неправильный формат количества. Должно быть целое число");
                        continue;
                    }
                };
                let address = parts[3];
                
                warehouse.remove(product, quantity, address);
            },
            "INFO" => {
                warehouse.info();
            },
            "EXIT" => {
                break;
            },
            _ => {
                println!("Ошибка: Неизвестная команда '{}'. Доступные команды: ADD, REMOVE, INFO, EXIT", parts[0]);
            }
        }
    }
}

File: Task1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <iomanip>
#include <algorithm>
#include <limits>
using namespace std;

class Warehouse {
private:
    struct Cell {
        string address;
        string product;
        int quantity;
    };

    int zones;
    int shelves_per_zone;
    int sections_per_shelf;
    int shelves_per_section;
    int total_capacity;
    int used_capacity = 0;

    map<string, Cell> cells;

public:
    Warehouse(int z, int spz, int sec, int spl, int cap) 
        : zones(z), shelves_per_zone(spz), sections_per_shelf(sec), 
          shelves_per_section(spl), total_capacity(cap) {}

    bool Address(const string& address) {
        if (address.length() != 5) return false;
        
        char zone = address[0];
        if (zone < 'A' || zone >= 'A' + zones) return false;
        
        int shelf = stoi(address.substr(1, 2));
        if (shelf < 1 || shelf > shelves_per_zone) return false;
        
        int section = address[3] - '0';
        if (section < 1 || section > sections_per_shelf) return false;
        
        int shelf_num = address[4] - '0';
        if (shelf_num < 1 || shelf_num > shelves_per_section) return false;
        
        return true;
    }

    void ADD(const string& product, int quantity, const string& address) {
        if (!Address(address)) {
            cout << "Ошибка: Неверный адрес: " << address << endl;
            return;
        }

        if (quantity > 10) {
            cout << "Ошибка: Нельзя добавить более 10 единиц в ячейку" << endl;
            return;
        }

        if (cells.count(address)) {
            Cell& cell = cells[address];
            if (cell.product != product) {
                cout << "Ошибка: Ячейка " << address << " уже содержит " << cell.product << endl;
                return;
            }
            if (cell.quantity + quantity > 10) {
                cout << "Ошибка: Ячейка " << address << " не может содержать более 10 единиц (сейчас: " << cell.quantity << ")" << endl;
                return;
            }
            cell.quantity += quantity;
        } else {
            cells[address] = {address, product, quantity};
        }
        used_capacity += quantity;
        cout << "Добавлено " << quantity << " единиц " << product << " в " << address << endl;
    }

    void REMOVE(const string& product, int quantity, const string& address) {
        if (!Address(address)) {
            cout << "Ошибка: Неверный адрес: " << address << endl;
            return;
        }

        if (!cells.count(address)) {
            cout << "Ошибка: Ячейка " << address << " пуста" << endl;
            return;
        }

        Cell& cell = cells[address];
        if (cell.product != product) {
            cout << "Ошибка: Ячейка " << address << " содержит " << cell.product << ", а не " << product << endl;
            return;
        }

        if (cell.quantity < quantity) {
            cout << "Ошибка: Недостаточно " << product << " в ячейке " << address << " (доступно: " << cell.quantity << ")" << endl;
            return;
        }

        cell.quantity -= quantity;
        used_capacity -= quantity;
        cout << "Удалено " << quantity << " единиц " << product << " из " << address << endl;

        if (cell.quantity == 0) {
            cells.erase(address);
        }
    }

    void INFO() {
        double total_percent = (static_cast<double>(used_capacity) / total_capacity) * 100;
        cout << fixed << setprecision(1);
        cout << "Информация о складе:" << endl;
        cout << "Общая заполненность: " << total_percent << "%" << endl;

        map<char, int> zone_capacity;
        map<char, int> zone_used;
        
        for (char zone = 'A'; zone < 'A' + zones; ++zone) {
            zone_capacity[zone] = shelves_per_zone * sections_per_shelf * shelves_per_section * 10;
            zone_used[zone] = 0;
        }

        for (const auto& pair : cells) {
            char zone = pair.first[0];
            zone_used[zone] += pair.second.quantity;
        }

        for (char zone = 'A'; zone < 'A' + zones; ++zone) {
            double zone_percent = (static_cast<double>(zone_used[zone]) / zone_capacity[zone]) * 100;
            cout << "Зона " << zone << " заполнена на " << zone_percent << "%" << endl;
        }

        cout << "\nЗанятые ячейки:" << endl;
        for (const auto& pair : cells) {
            cout << pair.first << ": " << pair.second.product << " (" << pair.second.quantity << ")" << endl;
        }

        int total_cells = zones * shelves_per_zone * sections_per_shelf * shelves_per_section;
        int empty_cells = total_cells - cells.size();
        cout << "\nПустые ячейки: " << empty_cells << endl;
    }
};

int main() {
    Warehouse warehouse(1, 10, 7, 4, 2800);
    string command;
    while (true) {
        cout << ">>> ";
        if (!getline(cin, command)) {
            cin.clear();
            cout << "Ошибка: Неправильный ввод команды" << endl;
            continue;
        }

        if (command.empty()) {
            continue;
        }

        // Разбиваем команду на части
        size_t space1 = command.find(' ');
        string cmd = command.substr(0, space1);

        if (cmd == "ADD") {
            size_t space2 = command.find(' ', space1 + 1);
            if (space1 == string::npos || space2 == string::npos) {
                cout << "Ошибка: Неправильный формат команды ADD. Используйте: ADD <продукт> <количество> <адрес>" << endl;
                continue;
            }

            string product = command.substr(space1 + 1, space2 - space1 - 1);
            string rest = command.substr(space2 + 1);

            size_t space3 = rest.find(' ');
            if (space3 == string::npos) {
                cout << "Ошибка: Неправильный формат команды ADD. Используйте: ADD <продукт> <количество> <адрес>" << endl;
                continue;
            }

            string quantityStr = rest.substr(0, space3);
            string address = rest.substr(space3 + 1);

            try {
                int quantity = stoi(quantityStr);
                warehouse.ADD(product, quantity, address);
            } catch (...) {
                cout << "Ошибка: Неправильный формат количества. Должно быть целое число" << endl;
            }
        } 
        else if (cmd == "REMOVE") {
            size_t space2 = command.find(' ', space1 + 1);
            if (space1 == string::npos || space2 == string::npos) {
                cout << "Ошибка: Неправильный формат команды REMOVE. Используйте: REMOVE <продукт> <количество> <адрес>" << endl;
                continue;
            }

            string product = command.substr(space1 + 1, space2 - space1 - 1);
            string rest = command.substr(space2 + 1);

            size_t space3 = rest.find(' ');
            if (space3 == string::npos) {
                cout << "Ошибка: Неправильный формат команды REMOVE. Используйте: REMOVE <продукт> <количество> <адрес>" << endl;
                continue;
            }

            string quantityStr = rest.substr(0, space3);
            string address = rest.substr(space3 + 1);

            try {
                int quantity = stoi(quantityStr);
                warehouse.REMOVE(product, quantity, address);
            } catch (...) {
                cout << "Ошибка: Неправильный формат количества. Должно быть целое число" << endl;
            }
        }
        else if (cmd == "INFO") {
            warehouse.INFO();
        }
        else if (cmd == "EXIT") {
            break;
        }
        else {
            cout << "Ошибка: Неизвестная команда '" << cmd << "'. Доступные команды: ADD, REMOVE, INFO, EXIT" << endl;
        }
    }
    return 0;
}
